### PR TITLE
Feature/refactor cesm put data

### LINKED ISSUE #709: refactor cesm_put_data routine in CESM driver
Currently, the `cesm_put_data.c` routine in the CESM driver writes VIC output to the coupler fields for use by the coupler in RASM. However, this routine repeats many of the calculations that are already in `put_data.c` in `shared_all` in VIC, and thus this introduces potential bugs by allowing for writing different values to the history files and coupler fields. This needs to be updated to use the `out_data` array in `put_data`.

### PR DESCRIPTION
**Pull request by dgergel**
- [X] closes #709 
- [X] tests passed
- [ ] ~~new tests added~~
- [ ] ~~science test figures~~
- [X] ran uncrustify prior to final commit
- [X] ReleaseNotes entry

This PR refactors the `cesm_put_data` routine in the CESM driver so that the same output data is being written to the coupler fields and the history files, barring a few sign convention differences. It also sets the dust `l2x` fluxes, dry deposition velocity and MEGAN fluxes to 0, as I don't think we should be passing these with fill values to the coupler (previously in RASM we populated unused fields with zeros). 

A few remaining to-dos (which we may want to put in a separate PR): 
- updating calculation of aerodynamical resistances
- updating roughness and wind stresses (which depend on the aerodynamical resistances)

### CODE REVIEW

File: vic/drivers/cesm/src/cesm_put_data.c [hunk 1]
```diff
@@ -68,9 +68,12 @@ vic_cesm_put_data()
     energy_bal_struct          energy;
     snow_data_struct           snow;
     veg_var_struct             veg_var;
+    extern double           ***out_data;
```
Comment (jhamman) [inline vic/drivers/cesm/src/cesm_put_data.c]: put this above with the other `extern`s

File: vic/drivers/cesm/src/cesm_put_data.c [hunk 2]
```diff
@@ -68,9 +68,12 @@ vic_cesm_put_data()
...
 
     for (i = 0; i < local_domain.ncells_active; i++) {
         // Zero l2x vars (leave unused fields as MISSING values)
+	// NOTE: I think actually we want unused fields to be 
+	// filled in with zeros, that's what we did previously
```
Comment (jhamman) [inline vic/drivers/cesm/src/cesm_put_data.c]: Align comments. (Run uncrustify)

File: vic/drivers/cesm/src/cesm_put_data.c [hunk 3]
```diff
@@ -91,16 +94,91 @@ vic_cesm_put_data()
...
-        // l2x_vic[i].l2x_Fall_fco2_lnd = 0;
-        // l2x_vic[i].l2x_Fall_flxdst1 = 0;
-        // l2x_vic[i].l2x_Fall_flxdst2 = 0;
-        // l2x_vic[i].l2x_Fall_flxdst3 = 0;
-        // l2x_vic[i].l2x_Fall_flxdst4 = 0;
-        // l2x_vic[i].l2x_Fall_flxvoc = 0;
+        l2x_vic[i].l2x_Fall_fco2_lnd = 0;
+	// for now, uncommenting these out 
+	// because this might have been the source 
+	// of the floating point error in RFR runs
```
Comment (jhamman) [inline vic/drivers/cesm/src/cesm_put_data.c]: No problem uncommenting these. I don't think we need the comment here.

File: vic/drivers/cesm/src/cesm_put_data.c [hunk 4]
```diff
@@ -91,16 +94,91 @@ vic_cesm_put_data()
...
+	// that is already in the put_data routine
+
+	// temperature, VIC: K, CESM: K
+        l2x_vic[i].l2x_Sl_t = out_data[i][OUT_RAD_TEMP][0];
+
+        // albedo, VIC: fraction, CESM: fraction 
+	// Note: VIC does not partition its albedo, thus all types are
+	// the same value 
+        // force->NetShortAtmos net shortwave flux (+ down)
+        // SWup = force->shortwave[NR] - energy.NetShortAtmos
```
Comment (jhamman) [inline vic/drivers/cesm/src/cesm_put_data.c]: Do we need these last two comments?

File: vic/drivers/cesm/src/cesm_put_data.c [hunk 5]
```diff
@@ -91,16 +94,91 @@ vic_cesm_put_data()
...
+
+        // albedo: diffuse, visible
+        l2x_vic[i].l2x_Sl_avsdf = albedo;
+
+        // albedo: diffuse, near-ir
+        l2x_vic[i].l2x_Sl_anidf = albedo;
+
+        // snow height, VIC: cm, CESM: m
+	// convert to VIC units
+        l2x_vic[i].l2x_Sl_snowh = out_data[i][OUT_SNOW_DEPTH][0] / CM_PER_M;
```
Comment (jhamman) [inline vic/drivers/cesm/src/cesm_put_data.c]: are we sure CPL/WRF isn't expecting SWE?

File: vic/drivers/cesm/src/cesm_put_data.c [hunk 6]
```diff
@@ -91,16 +94,91 @@ vic_cesm_put_data()
...
+                                             out_data[i][OUT_LWNET][0]);
+
+	// turbulent heat fluxes
+        // latent heat, VIC: W/m2, CESM: W/m2
+        l2x_vic[i].l2x_Fall_lat = out_data[i][OUT_LATENT][0];
+
+        // sensible heat, VIC: W/m2, CESM: W/m2
+	// TO-DO: check sign in VIC 4 coupling, this is inconsistent with 
+	// the history file output sign 
+        l2x_vic[i].l2x_Fall_sen += -1 * out_data[i][OUT_SENSIBLE][0];
```
Comment (jhamman) [inline vic/drivers/cesm/src/cesm_put_data.c]: take care of this TODO as part of your next round of changes for this PR.

File: vic/drivers/cesm/src/cesm_put_data.c [hunk 7]
```diff
@@ -91,16 +94,91 @@ vic_cesm_put_data()
...
+
+        // evaporation, VIC: mm, CESM: kg m-2 s-1
+	// TO-DO should we incorporate bare soil evap?
+	// canopy corrections applied already 
+	// so not repeated here
+	evap = out_data[i][OUT_EVAP][0] * MM_PER_M;
+        l2x_vic[i].l2x_Fall_evap += -1 * evap /
+                                            global_param.dt;
```
Comment (jhamman) [inline vic/drivers/cesm/src/cesm_put_data.c]: do you really need this temporary `evap` variable?

File: vic/drivers/cesm/src/cesm_put_data.c (hunk 5 above)
Comment (dgergel) [inline vic/drivers/cesm/src/cesm_put_data.c]: In `~/wrf/phys/module_ra_cam.F`, it's defined as `Snow depth over land, water equivalent (m)`, which is confusing wording but I take it to mean that we should be returning SWE, not snow depth - so I think you're right

File: vic/drivers/cesm/src/cesm_put_data.c (hunk 6 above)
Comment (dgergel) [inline vic/drivers/cesm/src/cesm_put_data.c]: Yes - it should be negative. In VIC4 we took care of the sign convention in the `lnd_comp_mct.F90` routine, but I think we want to do this in the VIC driver now.

File: vic/drivers/cesm/src/cesm_put_data.c (hunk 7 above)
Comment (dgergel) [inline vic/drivers/cesm/src/cesm_put_data.c]: I don't think so - just took it out.

**Comment by dgergel** [comment]
@jhamman and @bartnijssen, this is ready for a full review. I'm holding off on making corresponding changes for PR #712 since they'll break the Travis tests, so I'll make those in a quick follow up PR after this and #712 have been merged.

File: vic/drivers/cesm/src/cesm_put_data.c [hunk 8]
```diff
@@ -91,14 +91,84 @@ vic_cesm_put_data()
...
+        l2x_vic[i].l2x_Sl_snowh = out_data[i][OUT_SWE][0] / MM_PER_M;
+
+        // net shortwave, VIC: W/m2, CESM: W/m2
+        l2x_vic[i].l2x_Fall_swnet = out_data[i][OUT_SWNET][0];
+
+        // longwave up, VIC: W/m2, CESM: W/m2
+        // adjust sign for CESM sign convention
+        l2x_vic[i].l2x_Fall_lwup = -1 *
+                                   (out_data[i][OUT_LWDOWN][0] -
+                                    out_data[i][OUT_LWNET][0]);
```
Comment (jhamman) [inline vic/drivers/cesm/src/cesm_put_data.c]: Do we have a longwave up variable in out_data?

Comment (dgergel) [inline vic/drivers/cesm/src/cesm_put_data.c]: Nope - we have `LWDOWN`, `LWNET` and the band-specific version of `LWNET` in `out_data`.

**Comment by dgergel** [comment]
@jhamman thanks. Just added a release notes entry.

**Comment by dgergel** [comment]
@bartnijssen @jhamman this should be ready to merge.

**Pull request merged**